src/Connector.AzureDataLake/PartitionedBuffer.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace CluedIn.Connector.AzureDataLake
+{
+    internal class PartitionedBuffer<TItem> : IDisposable
+    {
+        private readonly int _maxSize;
+        private readonly int _timeout;
+        private readonly Action<TItem[]> _bulkAction;
+        private readonly Dictionary<object, Buffer<TItem>> _buffers;
+
+        public PartitionedBuffer(int maxSize, int timeout, Action<TItem[]> bulkAction)
+        {
+            _maxSize = maxSize;
+            _timeout = timeout;
+            _bulkAction = bulkAction;
+            _buffers = new Dictionary<object, Buffer<TItem>>();
+        }
+
+        public async Task Add(TItem item, object partition)
+        {
+            Buffer<TItem> buffer;
+            lock (_buffers)
+            {
+                if (!_buffers.TryGetValue(partition, out buffer))
+                {
+                    _buffers.Add(partition, buffer = new Buffer<TItem>(_maxSize, _timeout, _bulkAction));
```

Comment:
Is it a good idea to have PartitionedBuffer accept `Action<object, TItem[]>` for `bulkAction` and we do something like `new Buffer<TItem>(_maxSize, _timeout, items => bulkAction(partition, items)));`?

Reply:
I did consider adding a TPartition and doing just that, but it didn't seem to add much value as the underlying Buffer<T> is unaware of the partition and still needs to grab the configuration from the first item